hal-x86_64/src/serial.rs

```diff
@@ -248,6 +278,39 @@ impl<'a> Lock<'a> {
 }
 
 impl<B> Lock<'_, B> {
+    pub fn check_interrupt_type(&mut self) -> io::Result<Option<Pc16550dInterrupt>> {
+        // IIR bits 0 through 3 describe what happened to produce an interrupt, with bits 4 and 5
+        // always 0, and bits 6 and 7 set to 1 if `FCR0=1`.
+        let iir = self.inner.inner.iir();
+
+        if iir & 0b0011_0000 != 0b0000_0000 {
+            return Err(io::Error::new(
+                io::ErrorKind::Other,
+                "IIR indicates bogus interrupt bits"
+            ));
+        }
+
+        // TODO(ixi): probably should check IIR bits 6 and 7? punting on everything related to FIFO
+        // though.
+
+        let interrupt = match self.inner.inner.iir() & 0b1111 {
+            0b0000 => Pc16550dInterrupt::ModemStatus,
+            0b0001 => { return Ok(None); }
+            0b0010 => Pc16550dInterrupt::TransmitterHoldingRegEmpty,
+            0b0100 => Pc16550dInterrupt::ReceivedDataAvailable,
+            0b0110 => Pc16550dInterrupt::ReceiverLineStatus,
+            0b1100 => Pc16550dInterrupt::CharacterTimeout,
```

Comment:
a thought: you could write this as `bits if bits == Pc16550dInterrupt::WhicheverVariant as u8` so that the bit patterns are unfuckupable...might compile worse though idk. hopefully this is a nop if it isn't == 0b0001...

Reply:
yeah... one of the enum crates gives you a nice to/from primitive function too right? i should just do that